common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -35,403 +35,210 @@ namespace {
-    const gko::batch_dense::BatchEntry<ValueType> &V_shared_entry,
-    const gko::batch_dense::BatchEntry<ValueType> &H_shared_entry,
-    const gko::batch_dense::BatchEntry<ValueType> &cs_shared_entry,
-    const gko::batch_dense::BatchEntry<ValueType> &sn_shared_entry,
-    const gko::batch_dense::BatchEntry<typename gko::remove_complex<ValueType>>
-        &rhs_norms_shared_entry,
-    const gko::batch_dense::BatchEntry<typename gko::remove_complex<ValueType>>
-        &res_norms_shared_entry)
+    group::thread_block_tile<config::warp_size> &warp_grp, const int restart,
+    const BatchMatrixType_entry &a_gl_entry,
+    const ValueType *const b_global_entry,
+    const ValueType *const x_global_entry, ValueType *const x_shared_entry,
+    ValueType *const r_shared_entry, const PrecType &prec_shared,
+    ValueType *const z_shared_entry, ValueType *const V_shared_entry,
+    ValueType *const H_shared_entry, ValueType *const cs_shared_entry,
+    ValueType *const sn_shared_entry,
+    typename gko::remove_complex<ValueType> &rhs_norms,
+    typename gko::remove_complex<ValueType> &res_norms)
 {
-    // Compute norms of rhs
-    compute_norm2<ValueType>(b_global_entry, rhs_norms_shared_entry);
-
-
-    // copy x from global to shared memory
-    copy(x_global_entry, x_shared_entry);
-
-    // r = b
-    copy(b_global_entry, r_shared_entry);
-
-
+    for (int i = threadIdx.x; i < a_gl_entry.num_rows; i += blockDim.x) {
+        x_shared_entry[i] = x_global_entry[i];
+        r_shared_entry[i] = b_global_entry[i];
+    }
     __syncthreads();
 
-
     // r = b - A*x
-    advanced_spmv_kernel(static_cast<ValueType>(-1.0), A_global_entry,
-                         gko::batch::to_const(x_shared_entry),
-                         static_cast<ValueType>(1.0), r_shared_entry);
+    single_advanced_spmv_kernel(static_cast<ValueType>(-1.0), a_gl_entry,
+                                x_shared_entry, static_cast<ValueType>(1.0),
+                                r_shared_entry);
     __syncthreads();
 
-    compute_norm2<ValueType>(gko::batch::to_const(r_shared_entry),
-                             res_norms_shared_entry);
-
-
-    // z = precond * r
-    prec_shared.apply(gko::batch::to_const(r_shared_entry), z_shared_entry);
-
-
-    for (int li = threadIdx.x;
-         li < H_shared_entry.num_rows * H_shared_entry.num_rhs;
-         li += blockDim.x) {
-        const int r = li / H_shared_entry.num_rhs;
-        const int c = li % H_shared_entry.num_rhs;
-
-        H_shared_entry.values[r * H_shared_entry.stride + c] =
-            zero<ValueType>();
+    if (threadIdx.x / config::warp_size == 0) {
+        single_compute_norm2(warp_grp, a_gl_entry.num_rows, b_global_entry,
+                             rhs_norms);
+    } else if (threadIdx.x / config::warp_size == 1) {
+        single_compute_norm2(warp_grp, a_gl_entry.num_rows, r_shared_entry,
+                             res_norms);
```

Comment:
You would probably need another `__syncthreads()` after this ?

Reply:
We don't need it for correctness, I think. Would we want it for performance? At the moment I don't think that's the case either.